src/components/TextField/TextField.vue

```diff
@@ -0,0 +1,405 @@
+  - but WITHOUT ANY WARRANTY; without even the implied warranty of
+  - MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  - GNU Affero General Public License for more details.
+  -
+  - You should have received a copy of the GNU Affero General Public License
+  - along with this program. If not, see <http://www.gnu.org/licenses/>.
+-->
+
+<docs>
+### Description
+
+General purpose text field component.
+Note: the inner html `input` element inherits all the attributes from the
+textField component so you can add things like `autocomplete`, `maxlength`
+and `minlength`.
+
+```
+<template>
+	<div class="wrapper">
+		<TextField :value.sync="text1"
+			label="Type something here"
+			:can-clear="text1 !== ''"
+			@clear="clearText">
+			<Magnify :size="16" />
+		</TextField>
+		<TextField :value.sync="text2"
+			label="Type something here"
+			:success="true"
+			@clear="clearText">
+		</TextField>
+		<TextField :value.sync="text3"
+			label="Type something here"
+			:label-visible="true"
+			@clear="clearText">
+			<Lock :size="16" />
+		</TextField>
+		<div class="external-label">
+			<label for="$refs.textField.id">External label</label>
+			<TextField :value.sync="text4"
+				ref="textField"
+				:label-visible="true"
+				:labelOutside= "true"
+				@clear="clearText" />
+		</div>
+	</div>
+</template>
+<script>
+import Magnify from 'vue-material-design-icons/Magnify'
+import Lock from 'vue-material-design-icons/Lock'
+
+export default {
+	data() {
+		return {
+			text1: '',
+			text2: '',
+			text3: '',
+			text4: '',
+		}
+	},
+
+	components: {
+		Magnify,
+		Lock,
+	},
+
+	methods: {
+		clearText() {
+			this.text1 = ''
+		}
+	}
+}
+</script>
+<style lang="scss" scoped>
+.wrapper {
+	display: flex;
+	gap: 4px;
+	align-items: flex-end;
+	flex-wrap: wrap;
+}
+
+.external-label {
+	display: flex;
+	width: 100%;
+}
+</style>
+```
+</docs>
+
+<template>
+	<div class="text-field">
+		<label v-if="!labelOutside && label !== undefined"
+			class="text-field__label"
+			:class="{ 'text-field__label--hidden': !labelVisible }"
+			:for="inputName">
+			{{ label }}
+		</label>
+		<div class="text-field__main-wrapper">
+			<input v-bind="$attrs"
+				ref="input"
+				:name="inputName"
+				class="text-field__input"
+				type="text"
+				:placeholder="computedPlaceholder"
+				aria-live="polite"
+				:class="{
+					'text-field__input--can-clear': canClear,
+					'text-field__input--leading-icon': hasLeadingIcon,
+					'text-field__input--trailing-icon': hasTrailingIcon,
+				}"
+				:value="value"
+				v-on="$listeners"
+				@input="handleInput">
+
+			<!-- Leading icon -->
+			<div class="text-field__icon text-field__icon--leading">
+				<!-- Leading material design icon in the text field, set the size to 18 -->
+				<slot />
+			</div>
+
+			<!-- Success icon -->
+			<div v-if="success" class="text-field__icon text-field__icon--trailing">
+				<Check :size="18" />
+			</div>
+
+			<!-- clear text button -->
+			<Button v-if="canClear"
```

Comment:
```suggestion
			<Button v-else-if="canClear"
```

Since they can't be used together make it obvious with if-else?